### DISCUSSION TITLE
Why does `ExecutionResult#getExtensions()` returns `Map<Object, Object>` while `addExtension()` only accepts a `String` key?

### DISCUSSION THREAD

**Discussion by victorcampos** [graphql-java/graphql-java#3729] [Q&A] [answered]
The title kinda says it all, I'm trying to understand the reasoning for it returning a `Map<Object, Object>` when it's known it can only be constructed with `String` keys.

This issue arose when creating a converter to protobuf and trying to feed `extensions` into a protobuf `Struct` since the key cannot be "anything", trying to wrap my head around why it's been designed this way. The lack of typing in GraphQL is quite challenging in a strongly typed language.

**Comment by bbakerman** [comment, marked as answer]
I think this is bad modelling in retrospect.

The graphql spec says nothing about what it should be

https://spec.graphql.org/October2021/#sec-Response-Format

> The response map may also contain an entry with key `extensions`. This entry, if set, must have a map as its value. This entry is reserved for implementors to extend the protocol however they see fit, and hence there are no additional restrictions on its contents.

It says it should be a map not what type of map.

I think we have modelled this too loosely.  In retrospection I think a `Map<String,Object>` would have been better since it all often becomes JSON anyway and needs keys to be strings in effect.

But nominally we are serialisation independent and in theory (but probably not in practice) some serialisation could support non string key maps say.  But this is a weak argument in reality

**Comment by victorcampos** [comment]
> It says it should be a map not what type of map.

I had the same interpretation of the spec, couldn't come up with any use case where the map would contain some other arbitrary type as key, as you mentioned (and probably also GraphQL's main serialisation format) it will become a JSON which requires keys to be strings.

Even though modelled too loosely I'd put the blame square on the spec by not being more precise, if I at least could find a use case where a key of any other type would be useful for the `extensions` I'd be more in peace...

Thanks a lot for the quick answer, I believe I will be safe when serialising to protobuf to assume the keys to always be strings for my usage :)
